022-feature-gates.md

```diff
@@ -0,0 +1,125 @@
+
+The following examples show how are these mechanisms used to give a better idea:
+
+* The `strimzi.io/use-connector-resources` annotation in the `KafkaConnect` custom resource defines whether the Connector Operator should be enabled or not for given Kafka Connect resource.
+* The `useServiceDnsDomain` in the listener configuration of the `Kafka` custom resource defines what kind of address should be used for given listener.
+  Whether fully qualified DNS name or not.
+* When the operator during Kafka upgrade sees that the old pods are still using the TLS sidecars, it will automatically loosen the TLS configuration to make sure the upgrade can be completed.
+
+## Motivation
+
+In many cases, the mechanisms described above work fine and are well suited.
+But there are also limitations:
+
+* Changes to the custom resources APIs have permanent character.
+  We can remove fields from the CRDs only when changing the API versions.
+  So adding fields to the Kafka CR is not a good option to gradually roll out changes or change a default behavior over multiple releases.
+* Sometimes, you also want to just change the behavior of the operator in general.
+  But you want to avoid having endless list of `enableFeatureX`, `disableFeatureY`, or `useZ` fields in the custom resources.
+* The use of the state of the existing Kubernetes resources works well for one-of operations.
+  That fits well for upgrades or for triggering actions such as restarting a pod.
+  But not for features which change the operator behavior during the whole run of the operator.
+
+This shows the there might be space for other ways how to enable or disable features.
+
+## Proposal
+
+Strimzi operators should support _feature gates_ to gate selected functionality and switch it on or off.
+Each feature gate will have a default state (enabled or disabled by default) and users will be able to manually enable or disable them.
+The feature gates will apply for the whole operator and alter its behavior.
+
+The feature gates should go through 3 different maturity stages:
+* Alpha (typically disabled by default)
+* Beta (typically enabled by default)
+* General Availability / GA (typically enabled by default or always enabled)
+
+Alpha stage indicates that the feature might be unstable (i.e. it might change in the future) or buggy.
+Beta stage indicates a that the feature is already well tested and most probably should not change.
+GA means that the feature is stable and should not change anymore.
+GA features might optionally also change to _always enabled features_.
+In this case the feature gate will be removed and it will not be possible to disable the feature anymore.
+The alpha and beta features might be removed in case they are deemed as not useful or not the solution.
+
+Typically, the gated features should be introduced using a Strimzi Proposal
+The proposal should also suggest the plan for how the feature will move through the different maturity stages.
```

Comment:
I suppose that wherever the feature flags are listed, we will comment/annotate the lines with their maturity level?
I fear if there is a split and the maturity is tracked via issues the knowledge may be harder to find/update

Reply:
I added a new sentence to specifically mention that it should be listed int he documentation.